### PR TITLE
VZ-10285: Add help for customizing MySQL my.cnf settings

### PR DESCRIPTION
**Pull request by mgianatagh**

### CODE REVIEW

File: content/en/docs/security/keycloak/keycloak.md [hunk 1]
```diff
@@ -1,16 +1,121 @@
 ---
 title: Customize Keycloak and MySQL
-description: Customize Verrazzano Keycloak and MySQL settings for high availability
+description: Customize Verrazzano Keycloak and MySQL settings
```
Comment (rosemarymarano) [inline content/en/docs/security/keycloak/keycloak.md:3]: I'd suggest removing the description test for this doc. It's somewhat redundant with the title and you have a head 3 directly following it. It appears like three headings in a row, which is a general no-no. (We're getting more discerning with regard to the description text, not always requiring it and definitely not using it where it's not needed.)

File: content/en/docs/security/keycloak/keycloak.md [hunk 2]
```diff
@@ -1,16 +1,121 @@
...
 aliases:
   - /docs/customize/keycloak
 ---
 
+### Customize MySQL settings for high availability
 * To scale the number of MySQL instances use the [Verrazzano Custom Resource]({{< relref "docs/reference/vpo-verrazzano-v1beta1.md" >}}), not the MySQL StatefulSet.
 Directly modifying the StatefulSet may change the status of the cluster to `ONLINE_PARTIAL`.
 * You must have at least one running `mysql-router` to access the MySQL [InnoDB Cluster](https://dev.mysql.com/doc/refman/8.0/en/mysql-innodb-cluster-introduction.html). Scaling the number of `mysql-router` instances to zero
 may result in the [MySQL Operator]({{< relref "docs/reference/vpo-verrazzano-v1beta1.md#install.verrazzano.io/v1beta1.MySQLOperatorComponent" >}}) permanently losing communication with the cluster and Keycloak being unable to communicate with MySQL.
 * There are limitations to MySQL group replication policy to provide distributed coordination between servers. See [MySQL Fault-tolerance](https://dev.mysql.com/doc/refman/8.0/en/group-replication-fault-tolerance.html).
```
Comment (rosemarymarano) [inline content/en/docs/security/keycloak/keycloak.md:15]: -> See MySQL [Fault-tolerance]

File: content/en/docs/security/keycloak/keycloak.md [hunk 3]
```diff
@@ -1,16 +1,121 @@
...
 
 For instructions to customize persistent storage settings, see [Customize Persistent Storage]({{< relref "docs/observability/storage.md " >}}).
+
+### Customize MySQL my.cnf settings
```
Comment (rosemarymarano) [inline content/en/docs/security/keycloak/keycloak.md:19]: -> MySQL `my.cnf` (code font) settings

File: content/en/docs/security/keycloak/keycloak.md [hunk 4]
```diff
@@ -1,16 +1,121 @@
...
-description: Customize Verrazzano Keycloak and MySQL settings for high availability
+description: Customize Verrazzano Keycloak and MySQL settings
 Weight: 9
 draft: false
 aliases:
   - /docs/customize/keycloak
 ---
 
+### Customize MySQL settings for high availability
 * To scale the number of MySQL instances use the [Verrazzano Custom Resource]({{< relref "docs/reference/vpo-verrazzano-v1beta1.md" >}}), not the MySQL StatefulSet.
```
Comment (rosemarymarano) [inline content/en/docs/security/keycloak/keycloak.md:11]: -> use the [Verrazzano custom resource]

File: content/en/docs/security/keycloak/keycloak.md [hunk 5]
```diff
@@ -1,16 +1,121 @@
...
 * You must have at least one running `mysql-router` to access the MySQL [InnoDB Cluster](https://dev.mysql.com/doc/refman/8.0/en/mysql-innodb-cluster-introduction.html). Scaling the number of `mysql-router` instances to zero
 may result in the [MySQL Operator]({{< relref "docs/reference/vpo-verrazzano-v1beta1.md#install.verrazzano.io/v1beta1.MySQLOperatorComponent" >}}) permanently losing communication with the cluster and Keycloak being unable to communicate with MySQL.
 * There are limitations to MySQL group replication policy to provide distributed coordination between servers. See [MySQL Fault-tolerance](https://dev.mysql.com/doc/refman/8.0/en/group-replication-fault-tolerance.html).
 
 For instructions to customize persistent storage settings, see [Customize Persistent Storage]({{< relref "docs/observability/storage.md " >}}).
+
+### Customize MySQL my.cnf settings
+The file my.cnf contains the main configuration for MySQL.  The contents of my.cnf can be customized by providing overrides to the Keycloak subcomponent MySQL in the Verrazzano custom resource.
+
+For example, the default value of `max_connections` can be overridden as follows:
```
Comment (rosemarymarano) [inline content/en/docs/security/keycloak/keycloak.md:22]: For example, you can override the default value of `max_connections` as follows:

File: content/en/docs/security/keycloak/keycloak.md [hunk 6]
```diff
@@ -1,16 +1,121 @@
...
+          - values:
+              serverConfig:
+                mycnf: |
+                  max_connections = 250
+```
+
+</div>
+{{< /clipboard >}}
+
+The MySQL Operator only supports my.cnf configuration overrides on an installation.  The following steps are required to make changes to the my.cnf configuration after the initial installation.
```
Comment (rosemarymarano) [inline content/en/docs/security/keycloak/keycloak.md:47]: -> The MySQL Operator supports `my.cnf` file configuration overrides only upon installation.
-> After initial installation, the following steps are required to make changes to the `my.cnf` file.

File: content/en/docs/security/keycloak/keycloak.md [hunk 7]
```diff
@@ -1,16 +1,121 @@
...
+1. Edit the Verrazzano custom resource and set the overrides for `serverConfig.mycnf` as shown above in the `max_connections` example.  For example:
```
Comment (rosemarymarano) [inline content/en/docs/security/keycloak/keycloak.md:48]: -> as shown previously in the `max_connections` example.

File: content/en/docs/security/keycloak/keycloak.md [hunk 8]
```diff
@@ -1,16 +1,121 @@
...
+{{< clipboard >}}
+<div class="highlight">
+
+```
+$ kubectl get innodbcluster -n keycloak mysql -o yaml
+```
+
+</div>
+{{< /clipboard >}}
+4. Edit the ConfigMap named `mysql-initconf` in namespace `keycloak` and update the settings in the `99-extra.cnf` section. For example:
```
Comment (rosemarymarano) [inline content/en/docs/security/keycloak/keycloak.md:78]: -> Edit the `mysql-initconf` ConfigMap in the `keycloak` namespace and update the settings

File: content/en/docs/security/keycloak/keycloak.md [hunk 9]
```diff
@@ -1,16 +1,121 @@
...
+  99-extra.cnf: |
+    # Additional user configurations taken from spec.mycnf in InnoDBCluster.
+    # Do not edit directly.
+    [mysqld]
+    max_connections = 250
+```
+
+</div>
+{{< /clipboard >}}
+5. Start a rollout restart of the MySQL StatefulSet. After the rollout restart completes the MySQL pods will be using the configuration overrides.
```
Comment (rosemarymarano) [inline content/en/docs/security/keycloak/keycloak.md:102]: -> After the rollout restart completes, (comma)

**Review by rosemarymarano — COMMENTED** [review]
Some edits and comments

File: content/en/docs/security/keycloak/keycloak.md (hunk 1 above)
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md]: Removed

File: content/en/docs/security/keycloak/keycloak.md (hunk 4 above)
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md]: fixed

File: content/en/docs/security/keycloak/keycloak.md (hunk 2 above)
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md]: fixed

File: content/en/docs/security/keycloak/keycloak.md (hunk 3 above)
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md]: fixed

File: content/en/docs/security/keycloak/keycloak.md [hunk 10]
```diff
@@ -1,16 +1,121 @@
...
 Directly modifying the StatefulSet may change the status of the cluster to `ONLINE_PARTIAL`.
 * You must have at least one running `mysql-router` to access the MySQL [InnoDB Cluster](https://dev.mysql.com/doc/refman/8.0/en/mysql-innodb-cluster-introduction.html). Scaling the number of `mysql-router` instances to zero
 may result in the [MySQL Operator]({{< relref "docs/reference/vpo-verrazzano-v1beta1.md#install.verrazzano.io/v1beta1.MySQLOperatorComponent" >}}) permanently losing communication with the cluster and Keycloak being unable to communicate with MySQL.
 * There are limitations to MySQL group replication policy to provide distributed coordination between servers. See [MySQL Fault-tolerance](https://dev.mysql.com/doc/refman/8.0/en/group-replication-fault-tolerance.html).
 
 For instructions to customize persistent storage settings, see [Customize Persistent Storage]({{< relref "docs/observability/storage.md " >}}).
+
+### Customize MySQL my.cnf settings
+The file my.cnf contains the main configuration for MySQL.  The contents of my.cnf can be customized by providing overrides to the Keycloak subcomponent MySQL in the Verrazzano custom resource.
```
[the remark this answers is not in the document]
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md]: fixed

File: content/en/docs/security/keycloak/keycloak.md (hunk 5 above)
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md]: fixed

File: content/en/docs/security/keycloak/keycloak.md (hunk 6 above)
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md]: fixed

File: content/en/docs/security/keycloak/keycloak.md (hunk 7 above)
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md]: fixed

File: content/en/docs/security/keycloak/keycloak.md [hunk 11]
```diff
@@ -1,16 +1,121 @@
...
+{{< clipboard >}}
+<div class="highlight">
+
+```
+$ kubectl -n verrazzano-install rollout status deployment/verrazzano-platform-operator
+```
+
+</div>
+{{< /clipboard >}}
+3. The MySQL InnoDBCluster object is updated by the Verrazzano platform-operator to contain the `serverConfig.mycnf` overrides.  The following command can be used to view the contents of the InnoDBCluster object.
```
[the remark this answers is not in the document]
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md]: fixed

File: content/en/docs/security/keycloak/keycloak.md (hunk 8 above)
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md]: fixed

File: content/en/docs/security/keycloak/keycloak.md [hunk 12]
```diff
@@ -1,16 +1,121 @@
...
+
+</div>
+{{< /clipboard >}}
+4. Edit the ConfigMap named `mysql-initconf` in namespace `keycloak` and update the settings in the `99-extra.cnf` section. For example:
+{{< clipboard >}}
+<div class="highlight">
+
+```
+$ kubectl patch configmap -n keycloak mysql-initconf -p '{"data":{"99-extra.cnf": "# Additional user configurations taken from spec.mycnf in InnoDBCluster.\n# Do not edit directly.\n[mysqld]\nmax_connections = 250\n"}}' --type=merge
+```
```
[the remark this answers is not in the document]
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md:85]: I split it up to a few lines, but the longest portion cannot be split.  Let me know if the split is better.

File: content/en/docs/security/keycloak/keycloak.md (hunk 9 above)
Comment (mgianatagh) [reply content/en/docs/security/keycloak/keycloak.md]: fixed

File: content/en/docs/security/keycloak/keycloak.md (hunk 12 above)
[the remark this answers is not in the document]
Comment (rosemarymarano) [reply content/en/docs/security/keycloak/keycloak.md:85]: Yep, it's better.

**Review by rosemarymarano — APPROVED** [review]
All good. Thanks.

**Pull request merged**